include/libcyphal/presentation/subscriber_impl.hpp

```diff
@@ -0,0 +1,289 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_PRESENTATION_SUBSCRIBER_IMPL_HPP_INCLUDED
+#define LIBCYPHAL_PRESENTATION_SUBSCRIBER_IMPL_HPP_INCLUDED
+
+#include "presentation_delegate.hpp"
+#include "shared_object.hpp"
+
+#include "libcyphal/common/cavl/cavl.hpp"
+#include "libcyphal/transport/msg_sessions.hpp"
+#include "libcyphal/transport/scattered_buffer.hpp"
+#include "libcyphal/transport/types.hpp"
+#include "libcyphal/types.hpp"
+
+#include <cetl/cetl.hpp>
+#include <cetl/pf17/cetlpf.hpp>
+
+#include <cstdint>
+#include <memory>
+#include <utility>
+
+namespace libcyphal
+{
+namespace presentation
+{
+
+/// Internal implementation details of the Presentation layer.
+/// Not supposed to be used directly by the users of the library.
+///
+namespace detail
+{
+
+// TODO: docs
+class SubscriberImpl final : public cavl::Node<SubscriberImpl>, public detail::SharedObject
+{
+public:
+    class CallbackNode : public cavl::Node<CallbackNode>
+    {
+    public:
+        struct Deserializer
+        {
+            struct Context;
+            using FunctionPtr = void (*)(Context&);
+
+            struct Context
+            {
+                cetl::pmr::memory_resource&         memory;
+                const TimePoint                     approx_now;
+                const transport::ScatteredBuffer&   buffer;
+                const transport::MessageRxMetadata& metadata;
+                CallbackNode*&                      next_node;
+
+            };  // Context
+
+            using TypeId = std::uintptr_t;
+            template <typename Message>
+            static TypeId getTypeId() noexcept
```

Comment:
This type id breaks in dynamically linked executables as covered in the design doc. Can we try the solution outlined in the design doc:

>one approach is to use the first instance of Subscriber to perform deserialization and then to deliver the deserialized message object to each of its siblings by value

Reply:
I believe that it won't like break-break. IMHO the worst what might happen is that de-serialization will be done more than once for such "cross" dyn-linked modules... I'm not sure even how to reproduce it actually, b/c Presentation object lives at some specific module, as well as its internally made `SubscriberImpl` instances (they are not supposed to travel somehow between module boundaries). Even user-facing `Subscriber<Msg>` types are again made by the Presentation, you can move them outside of module boundary, but still references its original impl object inside (which does actuall deserialization).

Reply:
I think you should be able to reproduce it as follows: Create a shared library that instantiates a subscriber of type T on a given Presentation instance. Create an executable that does the same. Load the shared library dynamically and request a subscriber from it. The executable should end up with two subscribers of the same type that rely on different deserializer.

The described scenario is not entirely hypothetical, you can easily imagine a complex application doing just that (there are ROS libraries out there that are structured similarly). I think the implications of unexpectedly doubling the execution time could be significant in some applications dealing with large messages (examples of these are given in the design doc).

I think it is acceptable to keep the current suboptimal implementation as is for now, but we should log this as a future improvement with a big TODO comment and a separate issue.

Reply:
I understand what you are describing in general (and why two incompatible types might end up with the same id - it was pretty well described also at the RTTI part of the design), but where I lost is...
>... Create an executable that does the same.

Will it be done on two **different** instances of the Presentation object?
- If yes, then I don't see how two subscribers, made on two different Presentations might collide...
- If no, then the way I see it is that actual "factory" of subscribers will stay on only one side... am I missing something?

Reply:
I've created issue #380 to address this problem.